src/riscv/lib/src/pvm/linux/signals.rs

```diff
@@ -123,7 +143,89 @@ impl Elem for LinuxSigAction {
         // field.
         unsafe {
             LinuxSigAction::read_from_bytes(from_raw_parts(source, Self::STORED_SIZE.get()))
-                .expect("Bad sigaction")
+                .expect("The signal action has the same layout as the ABI type")
+        }
+    }
+
+    unsafe fn write_unaligned(self, dest: *mut u8) {
+        // SAFETY: The bitwise representation is the same as `read_unaligned` and matches each
+        // field.
+        unsafe {
+            let _ = self.write_to(from_raw_parts_mut(dest, Self::STORED_SIZE.get()));
+        }
+    }
+}
+
+/// Linux siginfo_t, see <https://man7.org/linux/man-pages/man2/sigaction.2.html>
+///
+/// This struct only contains the fields that are currently used.
+#[repr(C)]
+#[derive(Immutable, IntoBytes, KnownLayout, TryFromBytes)]
+pub struct LinuxSigInfo {
+    pub si_signo: Signal,
+    si_errno: I32<LittleEndian>, // Unused in Linux
+    pub si_code: I32<LittleEndian>,
+}
+
+impl LinuxSigInfo {
+    pub fn new(si_signo: Signal, si_code: i32) -> Self {
+        Self {
+            si_signo,
+            si_errno: Default::default(),
+            si_code: si_code.into(),
+        }
+    }
+}
+
+const ALIGNED_SIGINFO_SIZE: u64 = size_of::<LinuxSigInfo>() as u64 + RISCV_ABI_SP_ALIGNMENT.get()
+    - (RISCV_ABI_SP_ALIGNMENT.get() % size_of::<LinuxSigInfo>() as u64);
+
+const ALIGNED_CONTEXT_SIZE: u64 = size_of::<LinuxUContext>() as u64 + RISCV_ABI_SP_ALIGNMENT.get()
+    - (RISCV_ABI_SP_ALIGNMENT.get() % size_of::<LinuxUContext>() as u64);
+
+const CONTEXT_STACK_SIZE: u64 = ALIGNED_SIGINFO_SIZE + ALIGNED_CONTEXT_SIZE;
+
+impl Elem for LinuxSigInfo {
```

Comment:
You implement a lot of these. Could you create a wrapper `ZeroCopyElem<T>` and implement `impl<T: FromBytes> Elem for ZeroCopyElem<T>` instead? This way you make the wrapper re-usable and don't have to implement them over and over again.